odoo/tools/expr_checker.py

```diff
@@ -0,0 +1,205 @@
+
+    def visit_Call(self, node):
+        node = self.generic_visit(node)
+
+        if not self.fncall:
+            raise Exception("safe_eval didn't permit you to call any functions")
+
+        return ast.Call(
+            func=ast.Name("__ast_check_fn", ctx=ast.Load()),
+            args=[node.func, ast.Name("__ast_check_type_fn", ctx=ast.Load())]
+            + node.args,
+            keywords=node.keywords,
+        )
+
+    def visit_Name(self, node):
+        node = self.generic_visit(node)
+
+        if node.id.startswith("_") and not self.priv:
+            raise NameError(f"safe_eval: didn't permit you to read private elements")
+
+        if node.id in self.reserved_name:
+            raise NameError(f"safe_eval: {node.id} is a reserved name")
+
+        return node
+
+    def visit_FunctionDef(self, node):
+        node = self.generic_visit(node)
+
+        if node.name in self.reserved_name:
+            raise NameError(f"safe_eval: {node.name} is a reserved name")
+
+        return node
+
+    def visit_Attribute(self, node):
+        node = self.generic_visit(node)
+
+        if isinstance(node.ctx, ast.Load):
+            return ast.Call(
+                func=ast.Name("__ast_check_attr_and_type", ctx=ast.Load()),
+                args=[node.value, ast.Constant(node.attr), node],
+                keywords=[],
+            )
+
+        elif isinstance(node.ctx, ast.Store):
+            raise ValueError(
+                "safe_eval: doesn't permit you to store values in attributes"
+            )
+
+        elif isinstance(node.ctx, ast.Del):
+            raise ValueError("safe_eval: doesn't permit you to delete attributes")
+
+
+def expr_checker(
+    expr,
+    get_attr,
+    allow_function_calls=True,
+    allow_private=False,
+    check_type=None,
+    check_function=None,
+    return_code=True,
+):
+    node_checker = NodeChecker(allow_function_calls, allow_private)
+    user_code = ast.unparse(node_checker.visit(ast.parse(expr)))
+
+    if return_code:
+        code = "\n".join(
+            [
+                cleandoc(
+                    getsource(get_attr).replace(get_attr.__name__, "__ast_check_attr")
+                ),
+                cleandoc(
+                    getsource(check_type).replace(
+                        check_type.__name__, "__ast_check_type_fn"
+                    )
+                ),
+                cleandoc(
+                    getsource(check_function).replace(
+                        check_function.__name__, "__ast_check_fn"
+                    )
+                ),
+                user_code,
+            ]
+        )
+
+    else:
+        code = user_code
+
+    def __ast_default_check_type(method, value):
+        if check_type is not None:
+            ret = check_type(method, value)
+            if ret is not None:
+                return ret
+
+        if (
+            type(value) not in __safe_type + __require_checks_type
+            or type(value) in __require_checks_type
```

Comment:
can't you just do this ?

```suggestion
            type(value) not in __safe_type

Reply:
It won't work because of the case of an attribute we check that we can read the attribute and then we check if the type is valid. 

In the case of `Good().tell_me_hi()` removing the `+ __require_checks_type` will raise an exception

Reply:
Look weird, the condition is
> x not in (a + b) OR x in b

Which is equivalent to
> (x not in a AND x not in B) OR x in b

> (condition_1 AND condition_2) or not condition_2

It should be the same as this no ?

> condition_1 or not condition_2

So
> type(value) not in __safe_type or type(value) not in __require_checks_type

(sorry if I miss something)

Reply:
I see it like that 

> If the type is not in either of the 2 whitelists (the classic one and the one that required checks) or it's in the one that requires check and it's in a return or in an argument context

Reply:
```python
in_whitelist = type(value) not in __safe_type + __require_checks_type
in_correct_ctx = type(value) in __require_checks_type and method in ["returned", "arguments"]
```

it the perfect condition the type should be:  `in_whitelist AND in_correct_ctx`

if we do the opposite we'll have: `not in_whitelist OR not in_correct_ctx`

Reply:
yes that's what I said, you can simplify the condition :) 
> type(value) not in __safe_type or type(value) not in __require_checks_type